sdk/include/opentelemetry/sdk/trace/exporter.h

```diff
@@ -0,0 +1,43 @@
+#pragma once
+
+#include "opentelemetry/nostd/shared_ptr.h"
+#include "opentelemetry/nostd/span.h"
+#include "opentelemetry/sdk/trace/span_data.h"
+
+OPENTELEMETRY_BEGIN_NAMESPACE
+namespace sdk
+{
+namespace trace
+{
+/**
+ * ExportResult is returned as result of exporting a span batch.
+ */
+enum class ExportResult
+{
+  // Batch was successfully exported.
+  rSuccess = 0,
+  // Exporting failed. The caller must not retry exporting the same batch; the
+  // batch must be dropped.
+  rFailedNotRetryable
+};
+/**
+ * SpanExporter defines the interface that protocol-specific span exporters must
+ * implement.
+ */
+class SpanExporter
+{
+public:
+  virtual ~SpanExporter() = default;
+
+  // Exports a batch of spans.
+  //
+  // This method must not be called concurrently for the same exporter
+  // instance.
+  virtual ExportResult Export(
+      nostd::span<nostd::shared_ptr<opentelemetry::sdk::trace::SpanData>> &spans) noexcept = 0;
```

Comment:
Can we use the Recordable class here and make SpanData one instance of it?

It's possible to build tracers that are much more efficient if they don't need to maintain accessor methods to their data like SpanData has, so I'd rather not have the interface require SpanData.

Reply:
Yes. When coming up with that, it wasn't clear to me optimized `Recordable`s are passed through the processor/exporter pipeline (I thought they end up in the `Recorder` and are processed there).

Having that cleared up, I can make `SpanData` an instance of `Recordable` and pass `const Recordable`s here.

Are you fine with a `Recordable` having a `GetSpanData` method? Which might be a no-op for optimized recordables? Otherwise exporter implementors would need to downcast each `Recordable` they get, which feels a bit cumbersome.

Reply:
What's wrong with using static_cast?

Have a method
```
std::unique_ptr<Recordable> MakeRecordable() noexcept {
    // Return a SpanData if you want here
}
``
Then you can
```
auto span_data = static_cast<SpanData*>(recordable);
```
And you can substitute SpanData for more efficient Recordables.